src/Servers/Kestrel/perf/Kestrel.Performance/SchedulerBenchmark.cs

```diff
@@ -0,0 +1,199 @@
+                    {
+                        _threadPoolScheduler.Schedule(_action, null);
+                    }
+                };
+        }
+
+        [Benchmark(OperationsPerInvoke = OperationsPerInvoke, Baseline = true)]
+        public void LockBasedIOQueue() => Schedule(_lockBasedAction);
+
+        [Benchmark(OperationsPerInvoke = OperationsPerInvoke)]
+        public void LockFreeIOQueue() => Schedule(_lockFreeAction);
+
+        [Benchmark(OperationsPerInvoke = OperationsPerInvoke)]
+        public void ThreadPoolDirect() => Schedule(_threadPoolAction);
+
+        private void Schedule(Action<int> scheduleAction) => Parallel.For(0, OuterLoopCount, scheduleAction);
+
+        public class IOQueueLockFree : PipeScheduler, IThreadPoolWorkItem
+        {
+            private readonly ConcurrentQueue<Work> _workItems = new ConcurrentQueue<Work>();
+            private int _doingWork;
+
+            public override void Schedule(Action<object> action, object state)
+            {
+                _workItems.Enqueue(new Work(action, state));
+
+                // Set working if it wasn't (via atomic Interlocked).
+                if (Interlocked.CompareExchange(ref _doingWork, 1, 0) == 0)
+                {
+                    // Wasn't working, schedule.
+                    System.Threading.ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
+                }
+            }
+
+            void IThreadPoolWorkItem.Execute()
+            {
+                while (true)
+                {
+                    while (_workItems.TryDequeue(out Work item))
+                    {
+                        item.Callback(item.State);
+                    }
+
+                    // All work done.
+
+                    // Set _doingWork (0 == false) prior to checking IsEmpty to catch any missed work in interim.
+                    // This doesn't need to be volatile due to the following barrier (i.e. it is volatile).
+                    _doingWork = 0;
+
+                    // Ensure _doingWork is written before IsEmpty is read.
+                    // As they are two different memory locations, we insert a barrier to guarantee ordering.
+                    Thread.MemoryBarrier();
+
+                    // Check if there is work to do
+                    if (_workItems.IsEmpty)
+                    {
+                        // Nothing to do, exit.
+                        break;
+                    }
+
+                    // Is work, can we set it as active again (via atomic Interlocked), prior to scheduling?
+                    if (Interlocked.Exchange(ref _doingWork, 1) == 1)
+                    {
+                        // Execute has been rescheduled already, exit.
+                        break;
+                    }
+
+                    // Is work, wasn't already scheduled so continue loop.
+                }
+            }
+        }
+
+        public class IOQueueLockBased : PipeScheduler, IThreadPoolWorkItem
```

Comment:
Thanks! I got similar results to you:

```
           Method |      Mean |     Error |    StdDev |         Op/s | Scaled | ScaledSD |   Gen 0 |   Gen 1 |  Allocated |
----------------- |----------:|----------:|----------:|-------------:|-------:|---------:|--------:|--------:|-----------:|
 LockBasedIOQueue |  33.39 ns | 0.6386 ns | 0.7098 ns | 29,950,473.3 |   1.00 |     0.00 |       - |       - |    1.93 KB |
  LockFreeIOQueue |  25.24 ns | 0.4880 ns | 0.4326 ns | 39,619,570.1 |   0.76 |     0.02 |       - |       - |    1.92 KB |
 ThreadPoolDirect | 223.03 ns | 1.9989 ns | 1.8697 ns |  4,483,751.5 |   6.68 |     0.15 | 31.2500 | 15.6250 | 2005.16 KB |
```

I think this was enough to merge this PR. I don't think we need to keep around IOQueueLockBased and merge it to master. Can we keep this micro-benchmark around but just test real schedulers like IOQueue, ThreadPool, Inline?

Reply:
Yep

Reply:
Done